scheduler/reconcile_util_test.go

```diff
@@ -791,37 +778,138 @@ func TestAllocSet_filterByTainted(t *testing.T) {
+			},
+			migrate:       allocSet{},
+			disconnecting: allocSet{},
+			reconnecting: allocSet{
+				"running-original": {
+					ID:            "running-original",
+					Name:          "web",
+					ClientStatus:  structs.AllocClientStatusRunning,
+					DesiredStatus: structs.AllocDesiredStatusRun,
+					Job:           testJob,
+					NodeID:        "disconnected",
+					TaskGroup:     "web",
+					AllocStates:   unknownAllocState,
+				},
+			},
+			ignore: allocSet{},
+			lost:   allocSet{},
+		},
+		{
+			// After an alloc is reconnected, it should be considered
+			// "untainted" instead of "reconnecting" to allow changes such as
+			// job updates to be applied properly.
+			name:                        "disco-client-reconnected-alloc-untainted",
+			supportsDisconnectedClients: true,
+			now:                         time.Now(),
+			taintedNodes:                nodes,
+			skipNilNodeTest:             false,
+			all: allocSet{
+				"running-reconnected": {
+					ID:            "running-reconnected",
+					Name:          "web",
+					ClientStatus:  structs.AllocClientStatusRunning,
+					DesiredStatus: structs.AllocDesiredStatusRun,
+					Job:           testJob,
+					NodeID:        "normal",
+					TaskGroup:     "web",
+					AllocStates:   reconnectedAllocState,
+				},
+			},
+			untainted: allocSet{
+				"running-reconnected": {
+					ID:            "running-reconnected",
+					Name:          "web",
+					ClientStatus:  structs.AllocClientStatusRunning,
+					DesiredStatus: structs.AllocDesiredStatusRun,
+					Job:           testJob,
+					NodeID:        "normal",
+					TaskGroup:     "web",
+					AllocStates:   reconnectedAllocState,
+				},
+			},
+			migrate:       allocSet{},
+			disconnecting: allocSet{},
+			reconnecting:  allocSet{},
+			ignore:        allocSet{},
+			lost:          allocSet{},
+		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
 			// With tainted nodes
 			untainted, migrate, lost, disconnecting, reconnecting, ignore := tc.all.filterByTainted(tc.taintedNodes, tc.supportsDisconnectedClients, tc.now)
-			require.Equal(t, tc.untainted, untainted, "with-nodes: %s", "untainted")
-			require.Equal(t, tc.migrate, migrate, "with-nodes: %s", "migrate")
-			require.Equal(t, tc.lost, lost, "with-nodes: %s", "lost")
-			require.Equal(t, tc.disconnecting, disconnecting, "with-nodes: %s", "disconnecting")
-			require.Equal(t, tc.reconnecting, reconnecting, "with-nodes: %s", "reconnecting")
-			require.Equal(t, tc.ignore, ignore, "with-nodes: %s", "ignore")
+			assert.Equal(t, tc.untainted, untainted, "with-nodes: %s", "untainted")
+			assert.Equal(t, tc.migrate, migrate, "with-nodes: %s", "migrate")
+			assert.Equal(t, tc.lost, lost, "with-nodes: %s", "lost")
+			assert.Equal(t, tc.disconnecting, disconnecting, "with-nodes: %s", "disconnecting")
+			assert.Equal(t, tc.reconnecting, reconnecting, "with-nodes: %s", "reconnecting")
+			assert.Equal(t, tc.ignore, ignore, "with-nodes: %s", "ignore")
```

Comment:
I switched to `assert` because these fields are all independent and it was annoying to find errors one by one 😅 . Now the test will print all failures at once.